### PR TITLE
feat: speed up restarts by not building topologies for terminated queries

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

fixes: https://github.com/confluentinc/ksql/issues/4923

Previously, on a restart, each ksqlDB node would build a query topology for each previously submitted persistent queries, i.e. CREATE TABLE AS SELECT, CREATE STREAM AS SELECT or INSERT INTO statement, _even if the associated query has already been terminated_.

Building the query topology takes a reasonable amount of time. By avoiding building these topologies for terminated queries the restart time will be much reduced for clusters with a lot of historic commands in the command topic.

A more complete solution to this problem is to move to a compacted topic for storing commands: https://github.com/confluentinc/ksql/pull/4659

### Testing done 

Usual.

### Reviewing notes:

I didn't take the approach suggested in #4923 by @rodesai :

> the fix should be simple - just store a Supplier<KafkaStreams> instead of KafkaStreams in QueryMetadata and only get it from start()

IMHO, such an approach would:
1. Complicate the `QueryMetadata` type unnecessarily. 
2. Be very hard to test that the topologies weren't being built, e.g. in `RecoveryTest`.
3. Brittle: a regression could easily be introduced if someone were to add code in the recovery processes that called any getting on `QueryMetadata` that caused the supplied to be invoked.

Hence, I went with a different approach: to compact the list of commands to be restored. This compaction searches through the list of commands, keeping track of any commands that start a persistent query. If it later finds a terminate for that query, it removes the terminate _and_ updates the originating command to no longer start a persistent query.

TBH, I don't really like either fix, but it serves a purpose in the short term until we have a better solution, i.e. https://github.com/confluentinc/ksql/pull/4659

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+   * <p>Finds any command's whose queries plans that are later terminated. Any such commands have
+   * their query plan and associated terminate command removed.
+   *
+   * <p>This compaction stops unnecessary creation of Streams topologies on a server restart.
+   * Building such topologies is relatively slow and best avoided.
+   *
+   * @param restoreCommands the list of commands to compact.
+   * @return the compacted list of commands.
+   */
+  static List<QueuedCommand> compact(final List<QueuedCommand> restoreCommands) {
```
Comment (agavra) [inline ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java:52]: this feels needlessly complicated (in place modification using the iterator is really hard for me to follow as I need to track which source collection is being modified and used where). Can we simplify it by having a first pass just collect the index of terminated queries and the terminate commands in `restoreCommands` and then have a second pass that just copies the ones we need (modifying the ones that need to be modified) into a new list?

That would also obviate the need for using `IdentityHashMap` as we're using the index in the list to uniquely identify a command.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  private static Set<QueuedCommand> findTerminatedQueriesAndRemoveTerminateCommands(
+      final List<QueuedCommand> commands
+  ) {
+    final Map<QueryId, QueuedCommand> queries = new HashMap<>();
+    final Set<QueuedCommand> terminatedQueries = Collections.newSetFromMap(new IdentityHashMap<>());
+
+    final Iterator<QueuedCommand> it = commands.iterator();
+    while (it.hasNext()) {
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java:70]: Yeah, I know that all queued commands are now meant to be validated. However, I'd prefer to code this defensively. That way, if a bug is introduced in the code that means a duplicate or none-validated command is produced, this code won't just blow up.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java (hunk 1 above)
Comment (big-andy-coates) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java:52]: Meh, I don't really see this as being complicated and, given this is throw away work as we're moving to key compacted topic, do we really want to spend more time on it?  You're basically asking me to totally re-write the algo.

File: ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+   * their query plan and associated terminate command removed.
+   *
+   * <p>This compaction stops unnecessary creation of Streams topologies on a server restart.
+   * Building such topologies is relatively slow and best avoided.
+   *
+   * @param restoreCommands the list of commands to compact.
+   * @return the compacted list of commands.
+   */
+  static List<QueuedCommand> compact(final List<QueuedCommand> restoreCommands) {
+    final List<QueuedCommand> compacted = new ArrayList<>(restoreCommands);
```
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java:53]: done

**Comment by big-andy-coates** [comment]
@agavra @rodesai: One change made (minor).  Responded to your comments chaps - if you're happy with my response, then approve, otherwise, fight your corner ;) :p

**Pull request merged**